redfish_config: Add support to configure Redfish Host Interface

jyundt opened the pull request:
Adding another Manager command to redfish_config in order to set Redfish
Host Interface properties.

Fixes #3631

##### SUMMARY
This PR would allow users to configure Redfish Host Interface properties. As an example: toggle `InterfaceEnabled` to enable or disable the RF HI.

See #3631 for more details.


##### ISSUE TYPE
- Feature Pull Request


##### COMPONENT NAME
redfish_config
redfish_utils.py

##### ADDITIONAL INFORMATION
This command is mostly a copy/pasta of `SetManagerNic` command in `redfish_config`.

felixfontein commented on plugins/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -100,6 +100,12 @@
+    description:
+      - setting dict of HostInterface on OOB controller
+    type: dict
+    version_added: '3.9.0'
```
There will be no 3.9.0 release, the next release (on Tuesday) is 4.0.0.
```suggestion
    version_added: '4.0.0'
```

mraineri commented on plugins/module_utils/redfish_utils.py:
```diff
@@ -2792,3 +2792,81 @@ def set_manager_nic(self, nic_addr, nic_config):
+                a.get('@odata.id')]
+
+        # Fail if multiple Redfish Host Interfaces are detected
+        if len(uris) != 1:
```
In the vast majority of cases, there should only be one host interface on a manager, but there are designs we want to allow where one manager is managing multiple systems, thus allowing for multiple host interfaces. I think it would be good to have the option to allow a user to specify a host interface instance; if not provided, then simply go for the only interface.

Essentially:

```
if host_interface_id is specified and member exists:
    # Use exact interface
elif len(host_interface_list) == 1:
    # Use single host interface
else:
    # No matching interface; error
```

jyundt commented:
@mraineri as per your [previous comment](https://github.com/ansible-collections/community.general/pull/3632#discussion_r741020238), lemme know if aa72728 will work.

I tested with the following playbook snippet on a Supermicro server that had (only) one HostInterface resource (ID `1`).
```
---
- hosts: all
  gather_facts: false
  vars:
    bmc_username: root
    bmc_password: hunter2
    default_uri_timeout: 50
    default_uri_retries: 5
  tasks:
  - name: Gather initial BMC lan infomration
    command: ipmitool lan print
    register: ipmitool_output
    changed_when: ipmitool_output.rc != 0
    become: true

  - name: Set BMC IP address as fact
    set_fact:
      bmc_address: "{{ ipmitool_output.stdout | regex_search('IP Address\\s+:\\s+(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})', '\\1')
                                              | first }}"

  - name: Enable HostInterface without specifying a HostInterface resource ID
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: true
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - debug:
      var: redfish_results

  - name: Disable HostInterface for specified HostInterface resource ID
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: false
      hostinterface_id: "1"
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - name: Enable HostInterface for specified HostInterface resource ID that is not present
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: false
      hostinterface_id: "2"
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - debug:
      var: redfish_results
```

Output:
```
(2.10.5) [jyundt@aurora tests]$ ansible-galaxy collection install git@github.com:jyundt/community.general.git,add-redfish-hostinterface-config-support --force
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'community.general:4.0.0' to '/home/jyundt/.ansible/collections/ansible_collections/community/general'
Created collection for community.general at /home/jyundt/.ansible/collections/ansible_collections/community/general
community.general (4.0.0) was installed successfully
(2.10.5) [jyundt@aurora tests]$ cat rf_hi.yml 
---
- hosts: all
  gather_facts: false
  vars:
    bmc_username: root
    bmc_password: hunter2
    default_uri_timeout: 50
    default_uri_retries: 5
  tasks:
  - name: Gather initial BMC lan infomration
    command: ipmitool lan print
    register: ipmitool_output
    changed_when: ipmitool_output.rc != 0
    become: true

  - name: Set BMC IP address as fact
    set_fact:
      bmc_address: "{{ ipmitool_output.stdout | regex_search('IP Address\\s+:\\s+(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})', '\\1')
                                              | first }}"

  - name: Enable HostInterface without specifying a HostInterface resource ID
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: true
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - debug:
      var: redfish_results

  - name: Disable HostInterface for specified HostInterface resource ID
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: false
      hostinterface_id: "1"
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - name: Enable HostInterface for specified HostInterface resource ID that is not present
    community.general.redfish_config:
      category: Manager
      command: SetHostInterface 
      hostinterface_config:
        InterfaceEnabled: false
      hostinterface_id: "2"
      baseuri: "{{ bmc_address }}"
      username: "{{ bmc_username }}"
      password: "{{ bmc_password }}"
      timeout: "{{ default_uri_timeout }}"
    retries: "{{ default_uri_retries }}"
    register: redfish_results

  - debug:
      var: redfish_results

(2.10.5) [jyundt@aurora tests]$  ansible-playbook  rf_hi.yml -i rf_hi_hosts.yml -l smc_cloud_dc

PLAY [all] *********************************************************************************************************************************

TASK [Gather initial BMC lan infomration] **************************************************************************************************
ok: [smc_cloud_dc]

TASK [Set BMC IP address as fact] **********************************************************************************************************
ok: [smc_cloud_dc]

TASK [Enable HostInterface without specifying a HostInterface resource ID] *****************************************************************
changed: [smc_cloud_dc]

TASK [debug] *******************************************************************************************************************************
ok: [smc_cloud_dc] => {
    "redfish_results": {
        "changed": true,
        "failed": false,
        "msg": "Modified Host Interface"
    }
}

TASK [Disable HostInterface for specified HostInterface resource ID] ***********************************************************************
changed: [smc_cloud_dc]

TASK [Enable HostInterface for specified HostInterface resource ID that is not present] ****************************************************
fatal: [smc_cloud_dc]: FAILED! => {"changed": false, "msg": "HostInterface ID 2 not present."}

PLAY RECAP *********************************************************************************************************************************
smc_cloud_dc               : ok=5    changed=2    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

(2.10.5) [jyundt@aurora tests]$ 
```

russoz commented:
It would be great to take the opportunity and transpose those manual tests above to integration tests - even if they are marked `disabled`, they would still be there to be executed the next time by you or the next person.

felixfontein commented on plugins/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -100,6 +100,18 @@
+    description:
+      - setting dict of HostInterface on OOB controller
+    type: dict
+    version_added: '3.9.0'
```
```suggestion
    version_added: '4.1.0'
```
The next release accepting features will be 4.1.0 (same below).

felixfontein commented on plugins/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -100,6 +100,18 @@
+  hostinterface_config:
+    required: false
+    description:
+      - setting dict of HostInterface on OOB controller
```
```suggestion
      - Setting dict of HostInterface on OOB controller.
```
Descriptions should always have complete sentences.

felixfontein commented on plugins/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -100,6 +100,18 @@
+  hostinterface_id:
+    required: false
+    description:
+      - redfish HostInterface instance ID if multiple HostInterfaces are present
```
```suggestion
      - Redfish HostInterface instance ID if multiple HostInterfaces are present.
```

mraineri commented:
Pending the other nits mentioned by @felixfontein , this looks good to me. Will give the ship comment when addressed.

jyundt commented:
@mraineri 
> Pending the other nits mentioned by @felixfontein , this looks good to me. Will give the ship comment when addressed.

Added changes from @felixfontein.



@russoz 

> It would be great to take the opportunity and transpose those manual tests above to integration tests - even if they are marked disabled, they would still be there to be executed the next time by you or the next person.

I recognize I'm punting on tests, but because this requires hardware (and configuration) that is specific to my environment (and specific to this PR, I only have temporary access to the hardware used in the previous example), I don't know if there would be any value in adding integration tests for this.

felixfontein commented on plugins/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -100,6 +100,18 @@
+    description:
+      - Redfish HostInterface instance ID if multiple HostInterfaces are present.
+    type: str
+    version_added: '3.9.0'
```
```suggestion
    version_added: '4.1.0'
```
This one is still missing.

jyundt commented:
@felixfontein :facepalm: sorry, fixed in c391073.

jyundt commented:
@felixfontein fixed in b0c7538.

felixfontein reviewed (approved):
The formal aspects look good to me.

felixfontein commented:
@mraineri is this PR ok from your side as well?

mraineri commented:
shipit

russoz commented:
Other than the couple of idiomatic suggestions, LGTM

russoz commented:
@jyundt not for this PR, you might want to check if `ansible.module_utils.common.dict_transformations.recursive_diff()` might help you simplify the logic you added.

See that function's code at:
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/dict_transformations.py#L127

felixfontein commented:
@jyundt thanks for contributing this!
@mraineri @russoz thanks for reviewing!
